keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,462 @@
+- The mechanism must be extensible to future normalization behaviours which will be
+  required to support protobuf for CRs. But this KEP does not aim at defining these
+  and hence defining a custom equality which is compatible with normalization.
+- The mechanism is not supposed to allow different orders in lists to be considered equal.
+
+## Proposal
+
+We propose
+
+1. adding boolean vendor extensions to CRD OpenAPI schemas named `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` with `true` as the only valid value.
+2. do **strict deep-equal** comparison of those fields marked as immutable during
+   update validation in the request pipeline, with these exceptions:
+   - for fields with `x-kubernetes-immutable-keys: true` only the keys of the map or array (specified via `x-kubernetes-list-map-keys`) are compared.
+   - for fields with `x-kubernetes-list-type: set` the order is ignored.
+   
+   If that comparison fails, the request is rejected as 400 "Bad Request".
+   
+We create another KEP to define custom normalization steps for CRs done during 
+deserialization from etcd and when receiving a request (just after pruning and defaulting).
+
+### OpenAPI extension `x-kubernetes-immutable`
+
+The `x-kubernetes-immutable` vendor extension is set in the spec of fields,
+arrays and objects. It recursively applies to the whole subtree:
+
+```yaml
+properties:
+  foo:
+    type: array
+    items:
+      x-kubernetes-immutable: true
+      type: string
+```
+
+means that the `foo` array itself is mutable, but the items are not. This means that
+items can be deleted and added, but at each existing index the string values are
+immutable.
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+```
+
+means that the whole `.foo` object is immutable. It cannot be removed or set.
+
+If one wants to make only field immutable, but allow to delete an existing object
+or set it if unset, the following is used:
+
+```yaml
+properties:
+  foo:
+    type: object
+    properties:
+      "x":
+        type: string
+        x-kubernetes-immutable: true
+      "y":
+        type: string
+        x-kubernetes-immutable: true
+```
+
+for all fields `x` and `y` of `foo`.
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+    properties:
+      bar:
+        type: object
+        x-kubernetes-immutable: true
+```
+
+can be simplified to 
+
+```yaml
+properties:
+  foo:
+    type: object
+    x-kubernetes-immutable: true
+    properties:
+      bar:
+        type: object
+```
+
+At the root-level and inside `.metadata` of an object, `x-kubernetes-immutable` is forbidden.
+
+### OpenAPI extension `x-kubernetes-immutable-keys`
+
+The topology of arrays is specified via `x-kubernetes-list-type` and `x-kubernetes-list-map-keys`. We currently support the `atomic`, `set` and `map` types. These interact with immutability in the following way:
```

Comment:
technically, we can also have "atomic" maps so you might want to consider that.

Reply:
done. In theory we could allow key-immutability for those atomic maps, but as that would contradict the meaning that an atomic map is not a set, but one value. Changing one bit changes the whole value.